### ISSUE TITLE
[QST] Gather/Scatter in cute/cutlass 3

### ISSUE THREAD

**Issue by akamiru** [NVIDIA/cutlass#1330]
Hi everybody,

I'm currently trying to writing a trainer for a very small an oddly shaped network which requires a lot of gather/scatter. E.g. one layer looks like this:
  C = A0 * B0 + A1 * B1 ...

where the A matrices are 32x8 and the B matrices are 8xN. There are 36 possible A matrices so I could load them all to shared memory once. The B matrices need to be gathered along the col (six in total). So I probably need to gather them all using cp.async then do a single 32x48 * 48xN GEMM (F32F16F16F32) I guess? What's the best way to approach this in cute? Could this be done in cutlass?

Help would really be appreciated.

**Comment by thakkarV** [comment]
Sounds like you want a grouped gemm that supports gather/scatter? Have you taken a look at example 52 for inspiration? Happy to help with the design, but using CuTe is fairly natural with the extensions in Hopper gather/scatter GEMM

**Comment by akamiru** [comment]
Sadly I only have access to SM89 so I haven't looked into the Hooper examples too much. To be honest I'm still rather new to GPU computing and I'm kinda having a hard time understanding how cute/cutlass works. Especially the tiny sizes of my matrices don't seem to match the supported layouts.

Thinking about it a grouped GEMM with scatter/gather which runs on SM89 would pretty much allow me to do everything my trainer needs except for the first layer which is sparse. Would you be willing to discuss my network and help me make a plan how to tackle it via discord?

**Comment by thakkarV** [comment]
I'm on vacation for a couple weeks but can help asynchronously on this thread. The concepts presented in example 52 are applicable pretty much 1:1 over here. I'm specifically referring to how the gather tensors layouts are transcribed. You can use the same custom indexed gather strides used in that example for your usecase. I recommend starting with a single file cute kernel like the cute tutorial

**Comment by hwu36** [comment]
does https://github.com/NVIDIA/cutlass/blob/main/examples/24_gemm_grouped/gemm_grouped.cu meet your need to run multiple GEMMs in parallel?

does https://github.com/NVIDIA/cutlass/blob/main/examples/36_gather_scatter_fusion/gather_scatter_fusion.cu meet your need to fuse gather/scatter into a GEMM?

If yes to both, you could just merge ex36 into ex24 which is fairly easy.  gather/scatter adds these three [templates](https://github.com/NVIDIA/cutlass/blob/main/examples/36_gather_scatter_fusion/gather_scatter_fusion.cu#L258-L261) into the highest level templates.  and these three [arguments](https://github.com/NVIDIA/cutlass/blob/main/examples/36_gather_scatter_fusion/gather_scatter_fusion.cu#L359-L361) into top level arguments.  So, what you need to do is to add these into group gemm interface.

group gemm and gather/scatter use the same underlying gemm implementation.  so you just need to do some plumbing in the top levels device and kernel levels.

@shangz-ai

**Comment by jeromeku** [comment]
@hwu36 

I'm also interested in combining grouped gemms with gather / scatter fusion.  

I see that both examples use the same underlying kernel [GemmUniversal](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_universal.h#L65).  

Other than adapting the top-level arguments and templates per your previous answer, what other changes do I need to make to the underlying code in [GemmGrouped](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_grouped.h#L65) such that it properly constructs, initializes and passes the args for the individual gather / scatter kernels? 

I.e., other than instantiating the underlying `gemm` kernel for the grouped gemm per [above](https://github.com/NVIDIA/cutlass/blob/main/examples/36_gather_scatter_fusion/gather_scatter_fusion.cu#L258-L261) and adding the tensor indices to the grouped gemm [arguments struct](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_grouped.h#L130), what else needs to be done to pass these indices to the gather / scatter gemms?  What I'm not clear on is how the top-level [GemmGroup arguments](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/examples/24_gemm_grouped/gemm_grouped.cu#L1264) interact with the underlying kernels ([GemmUniversal](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/default_gemm_grouped.h#L214)).

**Comment by hwu36** [comment]
@jackkosaian ^ ^

**Comment by jackkosaian** [comment]
@jeromeku,

It sounds like you've already figured out where new `Arguments` should be placed: [here](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_grouped.h#L130).

You'll also need to add them to the kernel's `Params` struct [here](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_grouped.h#L217), similar to how they are added for `GemmUniversal::Params` [here](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_universal.h#L291-L293) (but noting that you'll have pointers to the gather/scatter index pointers for grouped GEMM since you want one list of indices per problem in the group).

Then, you need to augment the `kernel::GemmGrouped::operator()` to (1) determine which gather/scatter pointer to use for a given tile being processed by the thread block, and (2) pass these pointers down to lower levels of the kernel.

(1) can be achieved by indexing into the gather/scatter pointers in `Params` using `problem_idx` similar to what grouped GEMM does for A/B pointers [here](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_grouped.h#L357-L361).

(2) can be done by following the pattern of use of the gather/scatter indices that is used by `GemmUniversal::operator()` (e.g., [here](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_universal.h#L531-L537).)

I hope this helps!

**Comment by jeromeku** [comment]
@jackkosaian 

Thanks for the clear explanation!

For the `gather / scatter` kernels, if I'm gathering rows of `A` (and scattering into rows of `C / D`), are there additional checks that need to be implemented to ensure that each `gemm` size is valid?  

For example, if I'm using `TensorOps`, the minimum `M` is `16` as predetermined by the tensor core instruction and thus the warp and tile shapes need to be multiples of `16`.  

In the case that the gathered `M` doesn't meet this requirement, would [`can_implement`](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_universal.h#L365-L441) need to be changed to ensure this condition by, e.g., padding to a minimum block size?  Is this implemented in the codebase or are there related examples?

**Comment by jackkosaian** [comment]
Any sort of padding would need to be handled externally to `can_implement`. You would need to pad your tensors, problem shapes, etc. before setting them in the `Arguments` struct.

**Comment by jeromeku** [comment]
@jackkosaian 

After changing [`kernel::GemmGrouped::operator()`](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_grouped.h#L343-L473) to:
```cpp
 while (problem_visitor.next_tile())
          {

            GemmCoord problem_size = problem_visitor.problem_size();
            int32_t problem_idx = problem_visitor.problem_index();
            int32_t threadblock_idx = int32_t(problem_visitor.threadblock_idx());

            GemmCoord grid_shape = problem_visitor.grid_shape(problem_size);

            cutlass::gemm::GemmCoord threadblock_offset(
                int(threadblock_idx / grid_shape.n()) * Mma::Shape::kM,
                int(threadblock_idx % grid_shape.n()) * Mma::Shape::kN,
                0);

            // Load element pointers. Exchange pointers and strides if working on the transpose
            ElementA *ptr_A = reinterpret_cast<ElementA *>((kTransposed ? params.ptr_B[problem_idx] : params.ptr_A[problem_idx]));
            typename LayoutA::LongIndex ldm_A = (kTransposed ? params.ldb[problem_idx] : params.lda[problem_idx]);

            ElementB *ptr_B = reinterpret_cast<ElementB *>((kTransposed ? params.ptr_A[problem_idx] : params.ptr_B[problem_idx]));
            typename LayoutB::LongIndex ldm_B = (kTransposed ? params.lda[problem_idx] : params.ldb[problem_idx]);

            int *gather_indices_A = reinterpret_cast<int *>((kTransposed ? params.ptr_gather_B_indices[problem_idx] : params.ptr_gather_A_indices[problem_idx]));

            int *gather_indices_B = reinterpret_cast<int *>((kTransposed ? params.ptr_gather_A_indices[problem_idx] : params.ptr_gather_B_indices[problem_idx]));

            // Compute initial location in logical coordinates
            cutlass::MatrixCoord tb_offset_A{
                threadblock_offset.m(),
                0,
            };

            cutlass::MatrixCoord tb_offset_B{
                0,
                threadblock_offset.n()};

            // Compute position within threadblock
            int thread_idx = threadIdx.x;
            CUTLASS_TRACE_HOST("ldm_A " << ldm_A);
            CUTLASS_TRACE_HOST("LayouA(ldm_A) " << LayoutA(ldm_A));
            CUTLASS_TRACE_HOST("ldm_B " << ldm_B);
            CUTLASS_TRACE_HOST("LayouA(ldm_B) " << LayoutB(ldm_B));

            // Construct iterators to A and B operands
            typename Mma::IteratorA iterator_A(
                LayoutA(ldm_A),
                ptr_A,
                {problem_size.m(), problem_size.k()},
                thread_idx,
                tb_offset_A,
                gather_indices_A);

            typename Mma::IteratorB iterator_B(
                LayoutB(ldm_B),
                ptr_B,
                {problem_size.k(), problem_size.n()},
                thread_idx,
                tb_offset_B,
                gather_indices_B);

            typename Mma::FragmentC accumulators;

            accumulators.clear();

            // Broadcast the warp_id computed by lane 0 to ensure dependent code
            // is compiled as warp-uniform.
            int warp_idx = canonical_warp_idx_sync();

            int lane_idx = threadIdx.x % 32;

            //
            // Matrix multiply phase
            //

            // Construct thread-scoped matrix multiply
            Mma mma(shared_storage.kernel.main_loop, thread_idx, warp_idx, lane_idx);

            // Compute threadblock-scoped matrix multiply-add
            int gemm_k_iterations = (problem_size.k() + Mma::Shape::kK - 1) / Mma::Shape::kK;

            // Wait for all threads to finish their epilogue phases from the previous tile.
            __syncthreads();

            // Compute threadblock-scoped matrix multiply-add
            mma(
                gemm_k_iterations,
                accumulators,
                iterator_A,
                iterator_B,
                accumulators);

            //
            // Epilogue
            //

            EpilogueOutputOp output_op(params.output_op);

            ElementC *ptr_C = params.ptr_C[problem_idx];
            ElementC *ptr_D = params.ptr_D[problem_idx];

            LayoutC layout_C(params.ldc[problem_idx]);
            LayoutC layout_D(params.ldd[problem_idx]);

            int *gather_indices_D = params.ptr_scatter_D_indices[problem_idx];

            typename Epilogue::OutputTileIterator::Params params_C(layout_C);
            typename Epilogue::OutputTileIterator::Params params_D(layout_D);

            // Tile iterator loading from source tensor.
            // typename Epilogue::OutputTileIterator iterator_C(
            //     params_C,
            //     ptr_C,
            //     problem_size.mn(),
            //     thread_idx,
            //     threadblock_offset.mn());
            typename Epilogue::OutputTileIterator iterator_C(
                params_C,
                ptr_C,
                problem_size.mn(),
                thread_idx,
                threadblock_offset,
                gather_indices_D);

            // Tile iterator writing to destination tensor.
            // typename Epilogue::OutputTileIterator iterator_D(
            //     params_D,
            //     ptr_D,
            //     problem_size.mn(),
            //     thread_idx,
            //     threadblock_offset.mn());
            typename Epilogue::OutputTileIterator iterator_D(
                params_D,
                ptr_D,
                problem_size.mn(),
                thread_idx,
                threadblock_offset,
                gather_indices_D);

            Epilogue epilogue(
                shared_storage.kernel.epilogue,
                thread_idx,
                warp_idx,
                lane_idx);

            // Execute the epilogue operator to update the destination tensor.
            epilogue(
                output_op,
                iterator_D,
                accumulators,
                iterator_C);

            // Next tile
            problem_visitor.advance(gridDim.x);
          }
        }
      };
```
I'm getting this error when constructing the [`Epilogue::OutputTileIterator`](https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/kernel/gemm_grouped.h#L435-L454) for `C` and `D`:
```
 error: no instance of constructor "cutlass::epilogue::threadblock::PredicatedTileIterator<ThreadMap_, Element_, ScatterD, PermuteDLayout, UseCUDAStore>::PredicatedTileIterator [with ThreadMap_=cutlass::epilogue::threadblock::OutputTileOptimalThreadMap<cutlass::epilogue::threadblock::OutputTileShape<128, 8, 2, 1, 1>, cutlass::epilogue::threadblock::OutputTileShape<1, 8, 1, 1, 8>, 128, 8, 16>, Element_=cutlass::half_t, ScatterD=true, PermuteDLayout=cutlass::layout::NoPermute, UseCUDAStore=false]" matches the argument list
            argument types are: (cutlass::epilogue::threadblock::PredicatedTileIterator<cutlass::epilogue::threadblock::OutputTileOptimalThreadMap<cutlass::epilogue::threadblock::OutputTileShape<128, 8, 2, 1, 1>, cutlass::epilogue::threadblock::OutputTileShape<1, 8, 1, 1, 8>, 128, 8, 16>, cutlass::half_t, true, cutlass::layout::NoPermute, false>::Params, ElementC *, cutlass::Coord<2, int, int64_t>, int, cutlass::gemm::GemmCoord, int *)
                  params_C,
                  ^
```

**Comment by jeromeku** [comment]
@jackkosaian 

When using the base `gather / scatter` fusion kernel per `example 36`, how to properly configure the kernel to gather rows of `A` and scatter into rows of `C / D`?  

I changed `device::GemmUniversal` `Gather` [params](https://github.com/NVIDIA/cutlass/blob/bbe579a9e3beb6ea6626d9227ec32d0dae119a49/examples/36_gather_scatter_fusion/gather_scatter_fusion.cu#L259-L261) to `true, false, true` and the `gemm` arguments such that the `tensor_indices` are passed for `A` and `D` (and `nullptr` for `B`).  All layouts are set to `RowMajor`.

The results I'm getting are odd: for example, if I set `index_size = 64`, the first `64` values of the gathered rows of the computed result `D` are correct, but the rest are `0` (the initial value).  The non-gathered rows are 0, as expected.

**Comment by jeromeku** [comment]
Quick questions:

When filling a tensor with `BlockFillSequential` as such:
```cpp
cutlass::HostTensor<ElementInputA, LayoutInputA> tensor_a(
        problem_size.mk()); 
cutlass::reference::host::BlockFillSequential(
        tensor_a.host_data(), problem_size.m() * problem_size.k());
```
 For `ElementInputA` `cutlass::half_t` and `m = k = 128` this gives a sequence only up to `2048`, after which it repeats `2048` for the remaining values.

Guessing that I'm doing something wrong here...

**Comment by hwu36** [comment]
there is no 2049 in fp16.  you are doing correct thing.

**Comment by jeromeku** [comment]
Thanks @hwu36 

When I repeat the above using `host::TensorFillSequential`, I get an increasing sequence up to `16384` for `m = k = 128`  as expected, whereas for `BlockFillSequential`, the sequence stops at `2048` (and repeats `2048` thereafter).

**Comment by hwu36** [comment]
if you check the result returned by `host::TensorFillSequential`, you wont see 2049.

**Comment by hwu36** [comment]
`host::TensorFillSequential` uses slightly different way to calculate, but after 2048, they both are limited by fp16.

in fp16, 

2048+1=2048 <- `BlockFillSequentia`l way

fp16(2048 + 1)=2050 <- `TensorFillSequential` way

**Comment by jeromeku** [comment]
@jackkosaian

Are there any examples of `gather / scatter fusion` and `grouped_gemm` specifically for `Ampere` architectures using `Cutlass 3.0+` and `CuTe`?  

How would one implement the above (combining `gather / scatter` and  `grouped_gemm`) using the `3.0` API as opposed to the legacy `2.0` interface?

**Comment by jackkosaian** [comment]
> Are there any examples of gather / scatter fusion and grouped_gemm specifically for Ampere architectures using Cutlass 3.0+ and CuTe?

We do not have such an examples of this.

> How would one implement the above (combining gather / scatter and grouped_gemm) using the 3.0 API as opposed to the legacy 2.0 interface?

My suggestion would be to try to take a look at the CUTLASS 3 examples for [gather/scatter](https://github.com/NVIDIA/cutlass/tree/main/examples/52_hopper_gather_scatter_fusion) and [grouped GEMM](https://github.com/NVIDIA/cutlass/blob/main/examples/57_hopper_grouped_gemm/57_hopper_grouped_gemm.cu) (each of which currently target Hopper). You could consider adapting these to use SM80 CUTLASS 3 mainloops (similar to unit tests like [this one](https://github.com/NVIDIA/cutlass/blob/main/test/unit/gemm/device/sm80_gemm_f16_f16_f32_tensor_op_f32.cu)). Note, however, that GEMMs produced via the CUTLASS 3 API for CC < 90 are not currently as well optimized as those produced via the CUTLASS 2 API

**Comment by thakkarV** [comment]
> that GEMMs produced via the CUTLASS 3 API for CC < 90 are not currently as well optimized as those produced via the CUTLASS 2 API

That said, do not be discouraged. carefully crafted data and thread layouts can hit 95% peak perf on 3.x mainloop as well :)

**Comment by mnicely** [comment]
@jeromeku has you question been answered?

**Comment by jeromeku** [comment]
@jackkosaian 

How would I implement a `gather_scatter` with broadcasted scale factor in the epilogue?

More specifically:
- `gather` rows from `A`, `scatter` to `D`
- each scattered `D` row should be multiplied by its own scale factor, gathered from a vector of such factors

Seems like I'd have to customize `default_epilogue_with_broadcast` / `epilogue_with_broadcast`, `default_gemm_with_broadcast`, `gemm_universal_with_broadcast`, and `gemm_with_fused_epilogue` to accept `gather` / `scatter` indices?

**Comment by hwu36** [comment]
https://github.com/NVIDIA/cutlass/issues/1398#issuecomment-2008729784 from @apuaaChen can help you with the vectorized scaling part.  you still need to reuse the gather scatter logic in place.

**Comment by thakkarV** [comment]
Hi! We just released CUTLASS 3.5 and it contains an [example of CUTLASS 3.x based gather/scatter convolution kernel](https://github.com/NVIDIA/cutlass/tree/main/examples/59_ampere_gather_scatter_conv)

**Comment by jeromeku** [comment]
@mnicely yes, thanks.
